posthog/async_migrations/migrations/0004_replicated_schema.py

```diff
@@ -0,0 +1,305 @@
+    3. Update materialized views to write to `writable_TABLENAME`
+
+Constraints:
+
+    1. This migration relies on there being exactly one ClickHouse node when it's run.
+    2. For person and events tables, the schema tries to preserve any materialized columns.
+    3. This migration requires there to be no ongoing part merges while it's executing.
+    4. This migration depends on 0002_events_sample_by. If it didn't, this could be a normal migration.
+    5. This migration depends on the person_distinct_id2 async migration to have completed.
+"""
+
+
+@dataclass(frozen=True)
+class TableMigrationData:
+    name: str
+    new_table_engine: MergeTreeEngine
+    materialized_view_name: Optional[str]
+    create_materialized_view: Optional[str]
+
+    @property
+    def renamed_table_name(self):
+        return self.name
+
+    @property
+    def backup_table_name(self):
+        return f"{self.name}_backup_0004_replicated_schema"
+
+    @property
+    def tmp_table_name(self):
+        return f"{self.name}_tmp_0004_replicated_schema"
+
+
+@dataclass(frozen=True)
+class ShardedTableMigrationData(TableMigrationData):
+    rename_to: str
+    extra_tables: List[str]
+
+    @property
+    def renamed_table_name(self):
+        return self.rename_to
+
+
+class Migration(AsyncMigrationDefinition):
+
+    description = "Replace tables with replicated counterparts"
+
+    depends_on = "0003_fill_person_distinct_id2"
+
+    def is_required(self):
+        return "Distributed" not in self.get_current_engine("events")
+
+    @cached_property
+    def operations(self):
+        assert settings.CLICKHOUSE_REPLICATION, "CLICKHOUSE_REPLICATION env var needs to be set for this migration"
+        self.validate_number_of_nodes_in_cluster()
+
+        TABLE_MIGRATION_OPERATIONS = [
+            operation for table in self.tables_to_migrate() for operation in self.replicated_table_operations(table)
+        ]
+
+        return [
+            AsyncMigrationOperation.simple_op(sql="SYSTEM STOP MERGES", rollback="SYSTEM START MERGES"),
+            AsyncMigrationOperation(
+                fn=lambda _: setattr(config, "COMPUTE_MATERIALIZED_COLUMNS_ENABLED", False),
+                rollback_fn=lambda _: setattr(config, "COMPUTE_MATERIALIZED_COLUMNS_ENABLED", True),
+            ),
+            *TABLE_MIGRATION_OPERATIONS,
+            AsyncMigrationOperation(
+                fn=lambda _: setattr(config, "COMPUTE_MATERIALIZED_COLUMNS_ENABLED", False),
+                rollback_fn=lambda _: setattr(config, "COMPUTE_MATERIALIZED_COLUMNS_ENABLED", True),
+            ),
+            AsyncMigrationOperation.simple_op(sql="SYSTEM START MERGES", rollback="SYSTEM STOP MERGES",),
+        ]
+
+    def replicated_table_operations(self, table: TableMigrationData):
+        yield AsyncMigrationOperation.simple_op(
+            sql=f"""
+            CREATE TABLE {table.tmp_table_name} AS {table.name}
```

Comment:
All operations need to be resumable, so you need a if not exists here.

Reply:
Define resumability?

If you mean that it should be _idempotent_ - nope, it should not be. We need to ensure that the schemas of the two tables are _exactly_ identical, which the `AS` part of the query enforces. If the table exists but with a different schema, we have bigger problems.

Note that we drop the table on the rollback side of things. Though the issue around dropping tables/schema alterations still exists - I think we should roll back starting operation-1 not operation.